gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/oauth/OAuthViewModel.kt

```diff
@@ -0,0 +1,69 @@
+package com.gravatar.quickeditor.ui.oauth
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import androidx.lifecycle.viewmodel.CreationExtras
+import com.gravatar.quickeditor.data.service.WordPressOAuthService
+import com.gravatar.services.Result
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+internal class OAuthViewModel(
+    private val wordPressOAuthService: WordPressOAuthService,
+) : ViewModel() {
+    private val _uiState = MutableStateFlow(OAuthUiState())
+    val uiState: StateFlow<OAuthUiState> = _uiState.asStateFlow()
+
+    private val _actions = Channel<OAuthAction>(Channel.BUFFERED)
+    val actions = _actions.receiveAsFlow()
```

Comment:
> For the ViewModel itself, there's nothing extraordinary. One thing different than what we can be used to is that the Actions emitted from ViewModel are separate and not part of the UiState.

Why do you decide to have those `actions` outside the `UiState`? I would like to understand more about the reasons for that. I was reading a bit about it, and I'm not sure it's a good practice.

I know it's not the exact same case, but in [this post](https://medium.com/androiddevelopers/viewmodel-one-off-event-antipatterns-16a1da869b95) from Google, they describe some antipatterns that can apply here. 

I'm not saying this is wrong, but I would like to understand more why it's better to have it here.

Reply:
Here's a counterpoint for the article you mentioned - https://proandroiddev.com/viewmodel-events-as-state-are-an-antipattern-35ff4fbc6fb6.

When you do as Google suggests you end up with events/actions in the state that need to be [cleared once handled](https://developer.android.com/topic/architecture/ui-layer/events#consuming-trigger-updates). This feels like unnecessary overhead and complication. Not to mention that the State will grow quite a bit with more complicated navigation.

So my preference is to keep them separate.



BTW. I've decided to use `Action` name, to keep the `Event` name for UI events.

Reply:
Thanks for the article. Great discussion there! That makes everything more clear to me. 

> BTW. I've decided to use Action name, to keep the Event name for UI events.

👍